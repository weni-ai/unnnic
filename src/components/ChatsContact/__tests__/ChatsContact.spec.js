import { beforeEach, describe, expect, afterEach, test, vi } from 'vitest';
import { mount } from '@vue/test-utils';

import UnnnicChatsContact from '../ChatsContact.vue';
import moment from 'moment';

describe('UnnnicChatsContact', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(UnnnicChatsContact, {
      props: {
        title: 'Contact Name',
        lastMessage: { text: 'Message' },
      },
    });
  });

  afterEach(() => {
    // Reset moment locale after each test
    moment.locale('en');
  });

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should show hours:minutes when last interaction was today', async () => {
    const todayTime = moment().subtract(30, 'minutes').toISOString();

    await wrapper.setProps({ lastInteractionTime: todayTime });

    const interactionTimeText = wrapper.find(
      '[data-testid="last-interaction-time"]',
    );

    expect(interactionTimeText.exists()).toBe(true);
    expect(interactionTimeText.text()).toBe(moment(todayTime).format('HH:mm'));
  });

  it('should show "Yesterday" when last interaction was yesterday (English)', async () => {
    const yesterdayTime = moment()
      .subtract(1, 'day')
      .startOf('day')
      .add(12, 'hours')
      .toISOString();

    await wrapper.setProps({
      lastInteractionTime: yesterdayTime,
      locale: 'en',
    });

    const interactionTimeText = wrapper.find(
      '[data-testid="last-interaction-time"]',
    );

    expect(interactionTimeText.exists()).toBe(true);
    expect(interactionTimeText.text()).toBe('Yesterday');
  });

  it('should show "Ontem" when last interaction was yesterday (Portuguese)', async () => {
    const yesterdayTime = moment()
      .subtract(1, 'day')
      .startOf('day')
      .add(12, 'hours')
      .toISOString();

    await wrapper.setProps({
      lastInteractionTime: yesterdayTime,
      locale: 'pt-br',
    });

    const interactionTimeText = wrapper.find(
      '[data-testid="last-interaction-time"]',
    );

    expect(interactionTimeText.exists()).toBe(true);
    expect(interactionTimeText.text()).toBe('Ontem');
  });

  it('should show "Ayer" when last interaction was yesterday (Spanish)', async () => {
    const yesterdayTime = moment()
      .subtract(1, 'day')
      .startOf('day')
      .add(12, 'hours')
      .toISOString();

    await wrapper.setProps({
      lastInteractionTime: yesterdayTime,
      locale: 'es',
    });

    const interactionTimeText = wrapper.find(
      '[data-testid="last-interaction-time"]',
    );

    expect(interactionTimeText.exists()).toBe(true);
    expect(interactionTimeText.text()).toBe('Ayer');
  });

  it('should show date when last interaction was more than 1 day ago', async () => {
    const threeDaysAgoTime = moment().subtract(3, 'days').toISOString();

    await wrapper.setProps({ lastInteractionTime: threeDaysAgoTime });

    const interactionTimeText = wrapper.find(
      '[data-testid="last-interaction-time"]',
    );

    expect(interactionTimeText.exists()).toBe(true);
    expect(interactionTimeText.text()).toBe(
      moment(threeDaysAgoTime).format('L'),
    );
  });

  it('should show hours:minutes when difference is less than 1 hour', async () => {
    const recentTime = moment().subtract(30, 'minutes').toISOString();

    await wrapper.setProps({ lastInteractionTime: recentTime });

    const interactionTimeText = wrapper.find(
      '[data-testid="last-interaction-time"]',
    );

    expect(interactionTimeText.exists()).toBe(true);
    expect(interactionTimeText.text()).toBe(moment(recentTime).format('HH:mm'));
  });

  it('should show date when difference is more than 1 hour but not yesterday', async () => {
    const moreThanOneHourAgo = moment().subtract(2, 'hours').toISOString();

    await wrapper.setProps({ lastInteractionTime: moreThanOneHourAgo });

    const interactionTimeText = wrapper.find(
      '[data-testid="last-interaction-time"]',
    );

    expect(interactionTimeText.exists()).toBe(true);
    expect(interactionTimeText.text()).toBe(
      moment(moreThanOneHourAgo).format('HH:mm'),
    );
  });

  it('should not show time when lastInteractionTime is empty', async () => {
    await wrapper.setProps({ lastInteractionTime: '' });

    const interactionTimeText = wrapper.find(
      '[data-testid="last-interaction-time"]',
    );

    expect(interactionTimeText.exists()).toBe(false);
  });

  it('should default to English locale when no locale is provided', async () => {
    const yesterdayTime = moment()
      .subtract(1, 'day')
      .startOf('day')
      .add(12, 'hours')
      .toISOString();

    await wrapper.setProps({
      lastInteractionTime: yesterdayTime,
      // No locale prop provided
    });

    const interactionTimeText = wrapper.find(
      '[data-testid="last-interaction-time"]',
    );

    expect(interactionTimeText.exists()).toBe(true);
    expect(interactionTimeText.text()).toBe('Yesterday');
  });
});
