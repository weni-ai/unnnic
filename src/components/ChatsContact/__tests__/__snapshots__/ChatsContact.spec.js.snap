// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UnnnicChatsContact > should match snapshot 1`] = `
"<div data-v-8ce6efd5="" class="chats-contact" tabindex="0" style="position: relative; overflow: hidden;">
  <section data-v-c735998e="" data-v-8ce6efd5="" class="user-avatar user-avatar--xl"><span data-v-c735998e="">C</span></section>
  <div data-v-8ce6efd5="" class="chats-contact__infos">
    <h1 data-v-8ce6efd5="" class="chats-contact__infos__title ellipsis" title="Contact Name">Contact Name</h1>
    <div data-v-8ce6efd5="" class="chats-contact__infos__additional-information">
      <!--v-if-->
    </div>
  </div>
  <section data-v-8ce6efd5="" class="chats-contact__infos__unread-messages-container">
    <!--v-if-->
    <!--v-if-->
  </section>
  <transition-group-stub data-v-17112bed="" data-v-8ce6efd5="" name="grow" tag="div" appear="false" persisted="false" css="true" class="ripples"></transition-group-stub>
</div>"
`;
